symlink.1
.TH "SYMLINK" "1" "November 2015" "" ""
.SH "NAME"
\fBsymlink\fR
.P
Symlink \fBnode\fP repositories together as quickly as possible\.
.SH SYNOPSIS
.P
\fBsymlink [OPTION]\.\.\. repoDir\.\.\fP
.SH DESCRIPTION
.P
Links together all repositories using \fBnpm link\fP after having fetched the smallest amount of dependencies using \fBnpm install\fP\|\.
.P
Performs dependency analysis to ensure the least dependent modules are made available via \fBnpm link\fP first\.
.SH OPTIONS
.P
\fB\-g,\-\-global\fP    Globally installed module to be linked
.P
\fB\-e,\-\-execute\fP    Execute commands \- same as piping through \fBsh\fP
.SH EXAMPLES
.P
See how a set of repositories would be linked together
.P
\fBsymlink repos\fP
.P
Actually run the set of commands
.P
\fBsymlink repos | sh\fP
.P
Reuse large globally installed modules if you know what you are doing
.P
\fBsymlink repos \-g gulp\fP
.SH INSTALLATION
.P
Install globally through npm
.P
\fBnpm install \-g symlink\fP
.SH ALGORITHM
.P
There is an analysis step:
.RS 0
.IP \(bu 2
read \fBpackage\.json\fP of each module in the given directory/directories and collects \fBdependencies\fP + \fBdevDependencies\fP
.IP \(bu 2
finds the \fBlocal\fP ones (found in given directories)
.IP \(bu 2
finds the \fBexternal\fP ones (complement)
.IP \(bu 2
orders the modules by least inclusion

.RE
.P
Then the execution step for each module (by least inclusion):
.RS 0
.IP \(bu 2
\fBnpm link (localDeps) ∪ ((globals ∩ externalDeps))\fP
.IP \(bu 2
\fBnpm install (externalDeps ∖ globals)\fP
.IP \(bu 2
\fBnpm link\fP

.RE
.P
This ensures your repository directories are all linked together\.
.SH BUGS
.P
Please report bugs at \fIhttps://github\.com/clux/symlink/issues\fR


CHANGELOG.md
2.1.0 / 2015-11-16
=================
  * Add a man page
  * Add `.npmignore`

2.0.0 / 2015-03-28
=================
  * Dry run flag removed. Dry run is now the default.
  * Execute flag added `-e` or `--execute`.
  * Perform linking with either:
   1. `symlink repoDirs | sh`
   2. `symlink repoDirs --execute` (will log a little more)

1.1.0 / 2015-03-28
=================
  * Swap optimist for yargs and improve command line usage
  * Errors from child_process calls in CLI now result in non-zero return code

1.0.1 / 2015-03-28
==================
  * Errors from library use are now propagated correctly via callback

1.0.0 / 2015-03-26
==================
  * Allow multiple input directories to be listed before flags

0.3.0 / 2014-07-20
==================
  * `-r` flag is removed, now assumed the only unnamed argument
  * Rewrite so it can be used as a library:
    - package.json location is now fully asynchronous
    - library simply works out the commands like a -d dryrun
  * Unit test coverage

0.2.0 - 0.2.4 / 2013-10-07
==================
  * Add -g flag for global modules that can be linked in if present
  * Command output now perform multiple installs and links per line (shorter output)
  * Documentation improvements

0.1.1 / 2013-10-05
==================
  * Unit tests

0.1.0 / 2013-01-06
==================
  * Take a container directory rather than a bunch of sibling directories as input

0.0.3 / 2012-10-20
==================
  * documentation + code cleanup

0.0.1 / 2012-10-18
==================
  * initial release



lib/symlink.js
var join = require('path').join
  , $ = require('interlude')
  , fs = require('fs')
  , async = require('async');

var getJson = function (pth, cb) {
  var pkgjson = join(pth, 'package.json');
  fs.exists(pkgjson, function (exists) {
    if (exists) {
      fs.readFile(pkgjson, function (err, data) {
        cb(err, err ? null : { path: pth, data: JSON.parse(data) });
      });
    }
    else {
      cb(null, null); // no error but no package.json
    }
  });
};

var getJsons = function (dir, cb) {
  fs.readdir(dir, function (err, data) {
    if (err) {
      return cb(err);
    }
    var paths = data.map(function (str) {
      return join(dir, str);
    });
    async.map(paths, getJson, cb);
  });
};

var getJsonsFromDirectories = function (dirs, cb) {
  async.map(dirs, getJsons, cb);
};

var analyze = function (deps, absPaths, globals, names) {
  var ownDeps = {}      // deps in names
    , foreignDeps = {}; // deps not in names

  // partition dependencies
  Object.keys(deps).forEach(function (k) {
    ownDeps[k] = deps[k].filter($.elem(names));
    foreignDeps[k] = deps[k].filter($.notElem(names));
  });

  // sort deps in order of safe linking between each other
  var sorted = $.range(names.length).map(function () {
    var safe = $.firstBy(function (n) {
      // safe to link iff no local deps unlinked
      return !$.intersect(ownDeps[n], names).length;
    }, names);

    if (!safe) {
      // impossible to link a to b if b also tries to link to a without querying npm
      var err = "cannot link cyclically dependent: " + JSON.stringify(names);
      throw new Error(err);
    }
    names.splice(names.indexOf(safe), 1); // remove it from names
    return safe;
  });

  var cmds = [];
  sorted.forEach(function (n) {
    // then find all commands required for each module in the found safe order
    var cd = 'cd ' + absPaths[n] + ' && ';

    // npm link in -g requested modules and internal deps when they are specified
    var linked = $.intersect(globals, foreignDeps[n]).concat(ownDeps[n]);
    if (linked.length > 0) {
      cmds.push(cd + 'npm link ' + linked.join(' '));
    }

    // npm install remaining deps
    var remaining = foreignDeps[n].filter($.notElem(linked));
    if (remaining.length > 0) {
      cmds.push(cd + 'npm install ' + remaining.join(' '));
    }

    // npm link (to make this available to the modules with more dependencies)
    cmds.push(cd + 'npm link');
  });
  return cmds;
};


module.exports = function (dirs, globals, cb) {
  var deps = {}         // { module name -> [jsonDeps++jsonDevDeps] }
    , absPaths = {};    // { module name -> abs module path }

  getJsonsFromDirectories(dirs, function (err, datas) {
    if (err) {
      return cb(err);
    }
    var names = datas.reduce(function (acc, data) {
      var namesCurr = data.filter(function (o) {
        return o !== null; // folders with package.json
      }).map(function (o) {
        var json = o.data;
        var name = json.name;
        var mDeps = $.extend(json.dependencies || {}, json.devDependencies || {});
        deps[name] = Object.keys(mDeps);
        absPaths[name] = o.path;
        return name;
      });
      return acc.concat(namesCurr);
    }, []);

    var cmds;
    try {
      cmds = analyze(deps, absPaths, globals, names);
    }
    catch (err) {
      return cb(err);
    }
    return cb(null, cmds);
  });
};
